### PR TITLE
refactor: simplified config system

### LINKED ISSUE #5964: [bug]: `clip-vit-large-patch14` not found when converting a model, (CONVERT_MODEL_ROOT is set before the CLI args are applied)
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### Operating system

Windows

### GPU vendor

Nvidia (CUDA)

### GPU model

_No response_

### GPU VRAM

_No response_

### Version number

main 

### Browser

chrome

### Python dependencies

_No response_

### What happened

During model conversion of a safetensor I got an error 
```
File "d:\git\InvokeAI-test\.venv\lib\site-packages\transformers\utils\hub.py", line 462, in cached_file
    raise EnvironmentError(
OSError: Incorrect path_or_model_id: 'C:\Users\neil_\invokeai\models\core\convert\clip-vit-large-patch14'. Please provide either the path to a local folder or the repo_id of a model on the Hub.```

### What you expected to happen

I expected the clip-vit-large-patch14 model to be loaded from my configured models directory.

### How to reproduce the problem

start invoke with the CLI argument of --root set to a folder but don't have an invokeai.yaml in the current folder.   Then try and use a safetensor file that needs converting.

### Additional context

I have done some digging and belive I have tracked down what is causing this issue.  

In `invokeai\backend\model_manager\convert_ckpt_to_diffusers.py` ln:76 `CONVERT_MODEL_ROOT = InvokeAIAppConfig.get_config().models_path / "core/convert"` tries to use the config but the  parse_args have not been called yet so it retrieves the wrong path. 

### Discord username

skunkworxdark

### PR DESCRIPTION
**Pull request by psychedelicious**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x]  Feature
- [x] Bug Fix

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- Probably missed a few spots

## Description

I ran into some issues with interactions between argparse, omegaconf and pydantic validation + existing issues:
- argparse autogenerated CLI args for every config setting and this cannot handle structured config data (for example, the list of url regex & token pairs). It was not simple to just skip those settings due to the tight coupling of argparse and the config object.
- We were writing every default config setting to the config file. This is a longstanding issue. With this PR, that will no longer happen, and a simple migration from v3 config to v4 partially tidies all configs. I sneakily changed the name of `png_compress_level` to `pil_compress_level` - this will get everybody using the better default value of 1. 
- There was an awkward API with `InvokeAIAppConfig.get_config`.

It was all very intertwined and I couldn't figure out an easy fix, so I refactored the whole thing and simplified it substantially.

See the commit messages for details on the changes, especially 6b066f3b44af1cf079310c8b075abeea532e5899 which has the meat and potatoes. Most changes are just small changes to use the revised config system.

The API did change slightly from that commit, to support testing and the configurator.

## QA Instructions, Screenshots, Recordings

- Start up the app, your `invokeai.yaml` should change to the new flat format, and the old one backed up to `invokeai.yaml.bak`. Your YAML should now only contain settings that are different from defaults. Everything should work exactly the same as it did before - like custom paths for models, host, port.
- Restore your backed up YAML & delete the `.bak` file. Run the configurator - it should migrate the YAML as it starts and back it up, same as before. Change a couple settings and apply changes, the YAML should change accordingly.

It's hard to fully test this change because the config is used all over the place, but I was pretty darn careful when updating all the places the config is used.

## Merge Plan

This PR can be merged when approved

## Added/updated tests?

- [x] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

Closes #5964

### CODE REVIEW

**Comment by psychedelicious** [comment]
Please paste your original config.

The compress level is supposed to be reset and not be there. The new default is the best setting. For users who have the old default of 6 baked into their config, the new default of 1 represents up to 30% improvement every time we save an image.

**Comment by psychedelicious** [comment]
Looks like the default YAML file isn't being created. I'll figure that out tomorrow.

**Comment by hipsterusername** [comment]
```InvokeAI:
  Web Server:
    host: 127.0.0.1
    port: 9090
    allow_origins: []
    allow_credentials: true
    allow_methods:
    - '*'
    allow_headers:
    - '*'
  Features:
    esrgan: true
    internet_available: true
    log_tokenization: false
    patchmatch: true
  Memory/Performance:
    always_use_cpu: false
    free_gpu_mem: false
    max_cache_size: 12.0
    max_vram_cache_size: .25
    precision: auto
    sequential_guidance: false
    xformers_enabled: false
    tiled_decode: false
    lazy_offload: true
    log_memory_usage: false
    png_compress_level: 1
  Paths:
    autoimport_dir: autoimport
    lora_dir: autoimport\lora
    embedding_dir: autoimport\embedding
    controlnet_dir: autoimport\controlnet
    conf_path: configs\models.yaml
    models_dir: models
    legacy_conf_dir: configs\stable-diffusion
    db_dir: databases
    outdir: C:\Users\kentr\Invoke\outputs
    use_memory_db: false
  Logging:
    log_handlers:
    - console
    log_format: color
    log_level: debug```

**Comment by psychedelicious** [comment]
@hipsterusername Your resultant config file is almost exactly what it should be. Those values are all default values, they should not be in a config file that is meant to override defaults.

The things that are not correct are the cache sizes. Turns out you are using "legacy" config settings:
```yaml
    max_cache_size: 12.0 # This has been called "ram" for a while now
    max_vram_cache_size: .25 # This has been called "vram" for a while now
```

So what's needed to make your new yaml 100% correct is to add handling for these two settings into the simple migration logic. I'll fix that.

**Comment by blessedcoolant** [comment]
Works as intended on my end. A couple of small issues that I saw were some of the paths stayed in the new config even though they were the same as the defaults -- but the reason that might have happened is coz I was using absolute paths for those rather than relative.

For example "<fullpath>/models" was kept by the new config despite it being the exact same location as "<root>/models"

Rest of it seems good.

**Comment by blessedcoolant** [comment]
Is there anything holding this back?

**Comment by psychedelicious** [comment]
@blessedcoolant I'll be reviewing it again today, thinking through edge cases and testing interactions with the installer/configurator. Aiming to have this in RC3.

**Comment by psychedelicious** [comment]
I added two notable changes:

[feat: single app entrypoint with CLI arg parsing](https://github.com/invoke-ai/InvokeAI/pull/5920/commits/99c46ade272d5aaf89bffe11f28e5b673ee716b2)

We have two problems with how argparse is being utilized:
- We parse CLI args as the `api_app.py` file is read. This causes a problem pytest, which has an incompatible set of CLI args. Some tests import the FastAPI app, which triggers the config to parse CLI args, which receives the pytest args and fails.
- We've repeatedly had problems when something that uses the config is imported before the CLI args are parsed. When this happens, the root dir may not be set correctly, so we attempt to operate on incorrect paths.

To resolve these issues, we need to lift CLI arg parsing outside of the application code, but still let the application access the CLI args. We can create a external app entrypoint to do this.

- `InvokeAIArgs` is a simple helper class that parses CLI args and stores the result.
- `run_app()` is the new entrypoint. It first parses CLI args, then runs `invoke_api` to start the app.

The `invokeai-web` project script and `invokeai-web.py` dev script now call `run_app()` instead of `invoke_api()`.

The first time `get_config()` is called to get the singleton config object, it retrieves the args from `InvokeAIArgs`, sets the root dir if provided, then merges settings in from `invokeai.yaml`.

CLI arg parsing is now safely insulated from application code, but still accessible. And we don't need to worry about import order having an impact on anything, because by the time the app is running, we have already parsed CLI args. Whew!

[fix(install): resolve config-related issues with configurator](https://github.com/invoke-ai/InvokeAI/pull/5920/commits/c9ff982785ec8c3e7281dfeb143c61baa891b263)

- Do not use the singleton app config, this causes a lot of weirdness
- Update logic to use new config object
- Remove unused code

**Comment by psychedelicious** [comment]
I've spruced up the config docs a bit. 

Also fixed a problem where user-defined `models.yaml` paths weren't being migrated properly. This issue was brought to my attention in a [discord comment](https://discord.com/channels/1020123559063990373/1149506274971631688/1218125065342943412) about RC2. I noticed I had broken that migration logic in this PR and fixed it. I have _not_ fixed the issue on `main`.

**Comment by skunkworxdark** [comment]
I generally like the changes to the config, but I see a slight complication to how I currently dev my nodes.

Currently, I use the below launch.json for CLI arguments. This allows me to easily switch out to a separate DB and nodes folder but keep everything else from my main install. 

As most CLI arguments have been removed is there an easy way to achieve what I am currently doing?

part of my launch.json:
```json
  "version": "0.2.0",
  "configurations": [
    {
      // Run the InvokeAI backend & serve the pre-built UI
      "name": "InvokeAI Server",
      "type": "debugpy",
      "request": "launch",
      "program": "scripts/invokeai-web.py",
      "args": [
        // Your InvokeAI root dir (where `invokeai.yaml` lives)
        "--root",
        "d:\\ai\\invokeai",
        // Access the app from anywhere on your local network
        "--host",
        "0.0.0.0",
        // DB Folder
        "--db_dir",
        "d:\\git\\invoke-devdb",
        //Nodes Folder
        "--custom_nodes_dir",
        "d:\\git\\backup\\nodes-test",
        //MemoryDB
//        "--use_memory_db"
      ],
      "console": "integratedTerminal",
      "justMyCode": true
    },
```

**Comment by psychedelicious** [comment]
Fixed another edge case around custom models.yaml paths.

**Comment by psychedelicious** [comment]
@skunkworxdark I expected that at least the paths would need to be present. We can restore those CLI args, they just need to be manually implemented. 

(CLI args were autogenerated before - not easily being able to do that is the downside to the simplified config system)

**Comment by skunkworxdark** [comment]
Either that or have an option to read in an alternative config file name. Eg. Invokai-dev1.yaml. Then you retain the root folder but have changeable configs.  Not 100% if that is better but an option.

**Comment by lstein** [comment]
The `invokeai-model-install` CLI (which people still use) no longer accepts the `--root` argument:
```
invokeai-model-install --root ~/invokeai-test
[2024-03-16 20:19:02,372]::[InvokeAI]::INFO --> Patchmatch initialized
Traceback (most recent call last):
  File "/home/lstein/invokeai-main/.venv/bin/invokeai-model-install", line 8, in <module>
    sys.exit(main())
  File "/home/lstein/Projects/InvokeAI/invokeai/frontend/install/model_install.py", line 608, in main
    config.update_config(invoke_args)
  File "/home/lstein/Projects/InvokeAI/invokeai/app/services/config/config_default.py", line 191, in update_config
    new_config = self.model_validate(config)
  File "/home/lstein/invokeai-main/.venv/lib/python3.10/site-packages/pydantic/main.py", line 509, in model_validate
    return cls.__pydantic_validator__.validate_python(
pydantic_core._pydantic_core.ValidationError: 1 validation error for InvokeAIAppConfig
root
  Extra inputs are not permitted [type=extra_forbidden, input_value='/home/lstein/invokeai-test', input_type=str]
    For further information visit https://errors.pydantic.dev/2.6/v/extra_forbidden
```

**Comment by lstein** [comment]
> I generally like the changes to the config, but I see a slight complication to how I currently dev my nodes.
> 
> Currently, I use the below launch.json for CLI arguments. This allows me to easily switch out to a separate DB and nodes folder but keep everything else from my main install.
> 
> As most CLI arguments have been removed is there an easy way to achieve what I am currently doing?
> 
> Ideally, I would like to keep the same root folder but change some paths (DB, nodes, models etc...)
> 
> part of my launch.json:
> 
> ```json
>   "version": "0.2.0",
>   "configurations": [
>     {
>       // Run the InvokeAI backend & serve the pre-built UI
>       "name": "InvokeAI Server",
>       "type": "debugpy",
>       "request": "launch",
>       "program": "scripts/invokeai-web.py",
>       "args": [
>         // Your InvokeAI root dir (where `invokeai.yaml` lives)
>         "--root",
>         "d:\\ai\\invokeai",
>         // Access the app from anywhere on your local network
>         "--host",
>         "0.0.0.0",
>         // DB Folder
>         "--db_dir",
>         "d:\\git\\invoke-devdb",
>         //Nodes Folder
>         "--custom_nodes_dir",
>         "d:\\git\\backup\\nodes-test",
>         //MemoryDB
> //        "--use_memory_db"
>       ],
>       "console": "integratedTerminal",
>       "justMyCode": true
>     },
> ```

**Comment by lstein** [comment]
Oops!

**Comment by lstein** [comment]
> @psychedelicious Either that or have an option to read in an alternative config file name. Eg. Invokai-dev1.yaml. Then you retain the root folder but have changeable configs. Not 100% sure if that is better but would mean you could quickly test multiple configs by changing 1 cli param.

There is still the ability to change settings on the fly using environment variables, e.g.:
```
INVOKEAI_RAM=5 invokeai-web --root ~/invokeai-test/
```

**Comment by lstein** [comment]
`invokeai-configure` is not working 100% because the config object is getting created before the command line arguments are parsed. This means that `--root` is not honored on the command line. Partly this is due to the timing of config object creation in `backend/install/invokeai_configure.py`, and I have committed a fix for this. However, there seems to be another place where the singleton config object is being created unintentionally by `invokeai/app/invocations/fields.py` and I can't figure out how this is happening.

To illustrate the problem, open `invokeai/app/services/config/config_default.py` and add the following to `get_config()` on or around line 411 (after `args = InvokeAIArgs.args`):

```    
print(f'DEBUG: root={getattr(args, "root", None)}')
```

Now run this two-line script:

```
import invokeai.app.invocations.fields
print('foo')
```

I get the following output:

```
DEBUG: root=None
foo
```

The problem appears to be in `invokeai/app/invocations/__init__.py`, where `get_config()` is called in order to set a variable named `custom_nodes_path`. Can this call be moved to runtime rather than compile time?

I'm committing a workaround inside `invokeai_configure`.

**Comment by lstein** [comment]
`invokeai-configure` argument-parsing was broken (it had trouble finding the right root), but I updated it to use the isolated argument parsing pattern in `run_app()` and it seems to be OK now.

**Review by lstein — APPROVED** [review]
Looking pretty good. I found issues with a couple of the  CLI utilities not recognizing `--root` and fixed them.

**Comment by skunkworxdark** [comment]
> > @psychedelicious Either that or have an option to read in an alternative config file name. Eg. Invokai-dev1.yaml. Then you retain the root folder but have changeable configs. Not 100% sure if that is better but would mean you could quickly test multiple configs by changing 1 cli param.
> 
> There is still the ability to change settings on the fly using environment variables, e.g.:
> ```
> INVOKEAI_RAM=5 invokeai-web --root ~/invokeai-test/
> ```
> 

Can this be done from the launch.json in vscode?

**Comment by psychedelicious** [comment]
@skunkworxdark Yes, you can specify environment variables:
```json
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Invoke Server",
      "type": "debugpy",
      "request": "launch",
      "program": "scripts/invokeai-web.py",
      "args": ["--root", "/path/to/invokeai", "--host", "0.0.0.0"],
      "console": "integratedTerminal",
      "justMyCode": false,
      "env": {
        "INVOKEAI_RAM": "22.0"
      }
    },
}

File: invokeai/backend/install/check_root.py [hunk 1]
```diff
@@ -7,10 +7,15 @@
 from invokeai.app.services.config import InvokeAIAppConfig
 
 
+# TODO(psyche): Should this also check for things like ESRGAN models, database, etc?
```
Comment (brandonrising) [inline invokeai/backend/install/check_root.py:10]: I mean, can't you have your db and models dirs be in paths totally unrelated to the invokeai root? I think the messaging around this could be changed. Maybe we should just assert on whatever we think about the config would break the app and change the naming accordingly?

File: invokeai/app/services/config/config_default.py [hunk 2]
```diff
@@ -225,397 +47,383 @@ def validate_url_regex(cls, v: str) -> str:
...
-    @property
-    def profiles_path(self) -> Path:
-        """Path to the graph profiles directory."""
-        return self._resolve(self.profiles_dir)
 
-    @staticmethod
-    def find_root() -> Path:
-        """Choose the runtime root directory when not specified on command line or init file."""
-        return _find_root()
+@lru_cache(maxsize=1)
```
Comment (brandonrising) [inline invokeai/app/services/config/config_default.py:401]: This is such a neat way to managing singletons. Cool that it passes around the object instead of copies of it too. Wasn't sure if this would keep the changes on the second call but it did.
```
app_config = get_config()
app_config.log_level = "warning"
app_config = get_config()
```

File: invokeai/frontend/cli/arg_parser.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+- The currently active virtual environment's parent directory
+- `$HOME/invokeai`"""
+
+_ignore_missing_core_models_help = r"""If set, the app will ignore missing core diffusers conversion models.
+These are required to use checkpoint/safetensors models.
+If you only use diffusers models, you can safely enable this."""
+
+_parser = ArgumentParser(description="Invoke Studio", formatter_class=RawTextHelpFormatter)
+_parser.add_argument("--root", type=str, help=_root_help)
+_parser.add_argument("--ignore_missing_core_models", action="store_true", help=_ignore_missing_core_models_help)
```
Comment (brandonrising) [inline invokeai/frontend/cli/arg_parser.py:18]: Doesn't seem like this is getting set in the config

Comment (brandonrising) [reply invokeai/frontend/cli/arg_parser.py:18]: Adding an arg to pass a different source_path file to merge_from_file would probably get ahead of most of the issues faced by getting rid of the old args. Not sure how many people used them.

File: invokeai/backend/install/check_root.py (hunk 1 above)
Comment (psychedelicious) [reply invokeai/backend/install/check_root.py:10]: Renamed

File: invokeai/app/services/config/config_default.py (hunk 2 above)
Comment (psychedelicious) [reply invokeai/app/services/config/config_default.py:401]: Yeah, nifty trick :)

File: invokeai/frontend/cli/arg_parser.py (hunk 3 above)
Comment (psychedelicious) [reply invokeai/frontend/cli/arg_parser.py:18]: Whoops, fixed it not getting set

Comment (psychedelicious) [reply invokeai/frontend/cli/arg_parser.py:18]: > Adding an arg to pass a different source_path file to merge_from_file would probably get ahead of most of the issues faced by getting rid of the old args. Not sure how many people used them.

Yeah, was thinking about this, but I'm not sure how to make all the pieces work together. Can explore this in a followup.

**Review by brandonrising — APPROVED** [review]

**Comment by skunkworxdark** [comment]
Now that the Inkokeai.yaml will initially only contain the settings that are not default.  Will there be a method to output a full config including all the default settings. This could be used to help diagnose support issues. Obviously this should include a way to censor sensitive information like website download tokens.

**Comment by psychedelicious** [comment]
@skunkworxdark The docs here have all settings: https://invoke-ai.github.io/InvokeAI/features/CONFIGURATION/#all-settings

Maybe we could have the config file be written with settings commented out? I'm wary of putting all the settings there:
- Most people should only change a very small number of settings.
- If you accidentally leave a default manually set, and we determine a better default setting, users will be stuck on an old default.

**Comment by skunkworxdark** [comment]
> @skunkworxdark The docs here have all settings: https://invoke-ai.github.io/InvokeAI/features/CONFIGURATION/#all-settings
> 
> Maybe we could have the config file be written with settings commented out? I'm wary of putting all the settings there:
> - Most people should only change a very small number of settings.
> - If you accidentally leave a default manually set, and we determine a better default setting, users will be stuck on an old default.

I see all the settings there but not what the default will be if it is omitted from the yaml. I can find what they are by looking at the code but most users will not be able to do this. 

Commented defaults sounds like a nice compromise. But would they change if a default is updated.

Maybe a way of viewing a list of what the current settings are via the frontend including defaults that are not in the yaml file. As I said before would need sensitive info like tokens censored. Then you could keep the yaml clean and mark in the frontend list which are defaults and which are defaults but omitted etc..  could even add the descriptions that you linked in the docs. 

I'm not suggesting adding a way of editing the config just viewing them with defaults, but that might be nice in the long term.

**Comment by psychedelicious** [comment]
@skunkworxdark Ya I understand. An example config file will be written to `invokeai.example.yaml`:

```yaml
# This is an example file with default and example settings. Use the values here as a baseline.

# Internal metadata - do not edit:
schema_version: 4.0.0

# Put user settings here - see https://invoke-ai.github.io/InvokeAI/features/CONFIGURATION/:
host: 127.0.0.1
port: 9090
allow_origins: []
allow_credentials: true
allow_methods:
- '*'
allow_headers:
- '*'
log_tokenization: false
patchmatch: true
autoimport_dir: autoimport
models_dir: models
convert_cache_dir: models/.cache
legacy_conf_dir: configs
db_dir: databases
outputs_dir: outputs
custom_nodes_dir: nodes
log_handlers:
- console
log_format: color
log_level: info
log_sql: false
use_memory_db: false
dev_reload: false
profile_graphs: false
profiles_dir: profiles
ram: 7.5
vram: 0.25
convert_cache: 20.0
lazy_offload: true
log_memory_usage: false
device: auto
precision: auto
sequential_guidance: false
attention_type: auto
attention_slice_size: auto
force_tiled_decode: false
pil_compress_level: 1
max_queue_size: 10000
node_cache_size: 512
hashing_algorithm: blake3
remote_api_tokens:
- url_regex: cool-models.com
  token: my_secret_token
- url_regex: nifty-models.com
  token: some_other_token
```

I'll also be getting the JSON schema for the config put onto the [schema store](https://www.schemastore.org/json/). This will enable editors to get autocomplete and error highlighting in `invokeai.yaml`. VSCode supports this automatically, with no user configuration - I'm sure other editors do too.

**Comment by skunkworxdark** [comment]
I have tried using the env variables to configure some folders for dev however I have come across an issue.

When I use the below launch.json to start invoke for dev.  It writes out these env settings to the invokeai.yaml file overwriting the existing settings.   I wasn't expecting that.  As this is pointing to my normal invoke install this would mean that I would have to change the settings back to normal before I could run the normal installation. 

```json
  "configurations": [
    {
      // Run the InvokeAI backend & serve the pre-built UI
      "name": "InvokeAI Server New",
      "type": "debugpy",
      "request": "launch",
      "program": "scripts/invokeai-web.py",
      "console": "integratedTerminal",
      "justMyCode": true,
      "args": ["--root", "d:\\ai\\invokeai"],
      "env": {
        // Access the app from anywhere on your local network
        "INVOKEAI_HOST": "0.0.0.0",
        // DB Folder
        "INVOKEAI_DB_DIR": "d:\\git\\invoke-devdb",
        //Nodes Folder
        "INVOKEAI_CUSTOM_NODES_DIR": "d:\\git\\backup\\nodes-test",
        //MemoryDB
        //"INVOKEAI_USE_MEMORY_DB": "true"
        },
    },
```

**Comment by psychedelicious** [comment]
@skunkworxdark 
The example file (and everything I am describing here) is in a different, pending pr: https://github.com/invoke-ai/InvokeAI/pull/5991

That PR also includes the ability to use an arbitrary config file. So you can pass `--config /path/to/invokeai-dev.yaml` or whatever you'd like. Hopefully that's easier than using env vars and supports your use case. 

Re: env vars being written, I expect this to occur if you have them specified along with an uninitialized root directory. Not sure how to get around this, bc env vars are handled by the settings library and I don't see a way to defer merging them without disabling it completely.

**Comment by skunkworxdark** [comment]
> @skunkworxdark The example file (and everything I am describing here) is in a different, pending pr: #5991
> 
> That PR also includes the ability to use an arbitrary config file. So you can pass `--config /path/to/invokeai-dev.yaml` or whatever you'd like. Hopefully that's easier than using env vars and supports your use case.
> 
> Re: env vars being written, I expect this to occur if you have them specified along with an uninitialized root directory. Not sure how to get around this, bc env vars are handled by the settings library and I don't see a way to defer merging them without disabling it completely.

The `--config` option is great and does solve my use case issues.

What do you mean by `uninitialized root directory`?  If by that you mean you don't have a `schema_version: 4.0.0` style yaml then that is not what I am seeing.  With a 4.0 yaml file in place, I still get the env vars updating the yaml file.

Also, I am seeing the yaml file getting updated even if nothing has been changed inside the file every time I run my dev server.

I don't think either of these things is game-breaking but they are not how I would have expected them to work.  I would have thought that the env vars would be overrides to the config not updates to it. Also, I wouldn't have expected the config to be updated unless some config was changed within the app.

**Comment by psychedelicious** [comment]
@skunkworxdark By `uninitialized root directory`, I meant a root directory without a config file.

I'll figure out how to prevent env vars from being written to file, I agree that this is incorrect behaviour.
